### PR TITLE
Mixtral 8x7B support

### PR DESCRIPTION
**Pull request by pierrestock**
Adding support for `mistralai/Mixtral-8x7B-v0.1` and `mistralai/Mixtral-8x7B-Instruct-v0.1` models as described in [our blogpost](https://mistral.ai/news/mixtral-of-experts/). 

This is joint work between @zhuohan123, @WoosukKwon from the vLLM project and Mistral AI. 

It integrates fast sparse mixture of experts kernels from the [Megablocks](https://github.com/stanford-futuredata/megablocks/tree/main) project.

### CODE REVIEW

**Review by zhuohan123 — APPROVED** [review]
LGTM! Thank you for your contribution and the official support of vLLM from Mistral AI!!

File: vllm/model_executor/models/mixtral.py [hunk 1]
```diff
@@ -0,0 +1,534 @@
...
+                cache_event,
+            )
+        return hidden_states
+
+    def sample(
+        self,
+        hidden_states: Optional[torch.Tensor],
+        sampling_metadata: SamplingMetadata,
+    ) -> SamplerOutput:
+        hidden_states = self.norm(hidden_states)
```
Comment (WoosukKwon) [inline vllm/model_executor/models/mixtral.py:502]: nit: Can we do this in `forward`?

Comment (zhuohan123) [reply vllm/model_executor/models/mixtral.py:502]: Yes, I merged the PR asap from my phone to let people use the model. I think one other small thing is to add mixtral to the supported model list. I am AFK now, can you help fix this if possible?

**Comment by draganjovanovich** [comment]
I installed from latest main, installed stk, megablocks, latest flash_attn, transformers etc...
And got following error: 
```llm.engine.ray_utils.RayWorkerVllm object at 0x7f1657fe5bd0>)
  File "/home/ubuntu/mambaforge/lib/python3.10/site-packages/vllm/engine/ray_utils.py", line 32, in execute_method
    return executor(*args, **kwargs)
  File "/home/ubuntu/mambaforge/lib/python3.10/site-packages/torch/utils/_contextlib.py", line 115, in decorate_context
    return func(*args, **kwargs)
  File "/home/ubuntu/mambaforge/lib/python3.10/site-packages/vllm/worker/worker.py", line 88, in profile_num_available_blocks
    self.model_runner.profile_run()
  File "/home/ubuntu/mambaforge/lib/python3.10/site-packages/torch/utils/_contextlib.py", line 115, in decorate_context
    return func(*args, **kwargs)
  File "/home/ubuntu/mambaforge/lib/python3.10/site-packages/vllm/worker/model_runner.py", line 321, in profile_run
    self.execute_model(seqs, kv_caches)
  File "/home/ubuntu/mambaforge/lib/python3.10/site-packages/torch/utils/_contextlib.py", line 115, in decorate_context
    return func(*args, **kwargs)
  File "/home/ubuntu/mambaforge/lib/python3.10/site-packages/vllm/worker/model_runner.py", line 279, in execute_model
    hidden_states = self.model(
  File "/home/ubuntu/mambaforge/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1518, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
  File "/home/ubuntu/mambaforge/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1527, in _call_impl
    return forward_call(*args, **kwargs)
  File "/home/ubuntu/mambaforge/lib/python3.10/site-packages/vllm/model_executor/models/mixtral.py", line 488, in forward
    hidden_states = layer(
  File "/home/ubuntu/mambaforge/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1518, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
  File "/home/ubuntu/mambaforge/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1527, in _call_impl
    return forward_call(*args, **kwargs)
  File "/home/ubuntu/mambaforge/lib/python3.10/site-packages/vllm/model_executor/models/mixtral.py", line 439, in forward
    r = self.block_sparse_moe(self.ffn_norm(h))
  File "/home/ubuntu/mambaforge/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1518, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
  File "/home/ubuntu/mambaforge/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1527, in _call_impl
    return forward_call(*args, **kwargs)
  File "/home/ubuntu/mambaforge/lib/python3.10/site-packages/torch/utils/_contextlib.py", line 115, in decorate_context
    return func(*args, **kwargs)
  File "/home/ubuntu/mambaforge/lib/python3.10/site-packages/vllm/model_executor/models/mixtral.py", line 353, in forward
    x = ops.padded_gather(x, indices, bin_ids, bins, padded_bins,
  File "/home/ubuntu/mambaforge/lib/python3.10/site-packages/torch/autograd/function.py", line 539, in apply
    return super().apply(*args, **kwargs)  # type: ignore[misc]
  File "/home/ubuntu/mambaforge/lib/python3.10/site-packages/stk/backend/autocast.py", line 28, in decorate_fwd
    return fwd(*args, **kwargs)
TypeError: PaddedGatherOp.forward() takes 6 positional arguments but 7 were given
```

**Comment by WoosukKwon** [comment]
@draganjovanovich Thanks for reporting the error! Please re-install megablocks with `pip install git+https://github.com/stanford-futuredata/megablocks.git@v0.5.0`.

**Comment by tgale96** [comment]
Hi, it looks like the errors you're getting are from megablocks. Can you share more details on your environment? The latest issue looks like it may be because of the CUDA toolkit version you're using (maybe lets start a separate issue?).

**Comment by tgale96** [comment]
Hi Edson, you should be able to run Mixtral with by just installing MegaBlocks with `pip install megablocks` now. Could you give that a try? 

See related issues: 
https://github.com/vllm-project/vllm/issues/2017
https://github.com/vllm-project/vllm/issues/2022

**Pull request merged**